Store SSO User ID

GTcharles opened the pull request:
People want to track users within analytics tools across apps using the SSO unique ID that is assigned to users when they are created.

This PR adds an `external_id` attribute to store such information as well as changes the `Keycloak::SetupStudentAccount` to store said ID when creating a user account.

caws committed:
Add external_id attribute to Users.
The external ID attribute is used in other apps to store the keycloak unique id.
It is necessary to store it because otherwise we cannot identify users within the analytics tool (currently segment) across multiple apps.

caws committed:
Add id to attribute list.
KC should now return the ID too, as that is now stored in the User record.

caws committed:
Store the SSO unique ID when creating a user record.
Such ID is stored in the external ID attribute.
Also document that.

mpraglowski commented on lib/keycloak/fake_client.rb:
```diff
@@ -1,8 +1,10 @@
 module Keycloak
   class FakeClient
-    def initialize
+    def initialize(id: 'foo-bar-id')
```
Why FakeClient gets an id? Shouldn't it be an id generator?
Something like:
```
  def initialize(id: ->(email) { ...  })
  ...
  ```
  
  That will allow to have different ids for different users.

GTcharles replied:
The reason for allowing an ID to be set is because said ID [is verified in the test.](https://github.com/growthtribeacademy/pupilfirst/blob/store_sso_id/spec/lib/keycloak/setup_student_account_spec.rb)

mpraglowski replied:
I know - but instead of value you shall get an ID generator - and you should be able to generate different IDs for each created user. `SecureRandom.uuid` does that but will be PITA in tests when you will want to have know IDs.

GTcharles replied:
Yeah, I made it so that I can also mock the ID. It'll only resort to SecureRandom.uuid when nothing is passed.

mpraglowski commented on lib/keycloak/fake_client.rb:
```diff
@@ -1,8 +1,10 @@
+      @id = id
       reset!
     end
+    attr_accessor :id
```
Why we need that? Why public? Could be `private attr_reader :id`

GTcharles replied:
It's important to set this ID, as keycloak returns it for every user created.

Will move it to a private attribute.

mpraglowski commented on lib/keycloak/setup_student_account.rb:
```diff
@@ -15,6 +15,9 @@ def initialize(keycloak_client = Rails.configuration.keycloak_client)
       student = User.find(actor_id)
       create_keycloak_user(student.email, student.name)
+      sso_data = fetch_keycloak_user_data(student.email)
+      sso_id   = sso_data.fetch('id', nil)
```
Is this possible that there is no id in keycloak response? What then?
If not possible then we should raise error if there is no id.

GTcharles replied:
No, AFAIK the keycloak response always contains the user UUID.

mpraglowski replied:
So why we have `nil` as default? shouldn't we raise error here if there is no `id` ?

mpraglowski reviewed (commented):
I love the size of this PR - something I could quickly understand with clear & focused goal :)

caws committed:
Adjust id.
Move it to a private attribute.
ID can also be random uuid if not passed.

caws committed:
This is what I wanted.
Make it so that a proc is passed.
This way I can define for what users I want a specific ID or just leave it to define them randomly using SecureRandom.uuid.

caws committed:
Pass proc in test to specify ID for a single email.
This way it's possible to specify an id for a given email and leave other emails using random data.

caws committed:
Doh.
No need for this.

caws committed:
Make it optional.

GTcharles commented:
Hey, are we good to merge this?
